client/lib/url-search/index.js

```diff
@@ -2,37 +2,103 @@
  * External dependencies
  */
 import React from 'react';
+import debugFactory from 'debug';
+import page from 'page';
+import url from 'url';
+import { pick } from 'lodash';
 
 /**
  * Internal dependencies
  */
-import { doSearch, getSearchOpen } from 'lib/mixins/url-search';
+const debug = debugFactory( 'calypso:url-search' );
 
-export default function( Component ) {
-	const componentName = Component.displayName || Component.name || '';
+/**
+ * Function for constructing the url to page to. Here are some examples:
+ * 1. { uri: 'google.com', search:'hello' } --> 'google.com?s=hello'
+ * 2. {
+ *     uri: 'wordpress.com/read/search?q=reader+is+awesome',
+ *     search: 'reader is super awesome'
+ *     queryKey: 'q',
+ *    } --> 'wordpress.com/read/search?q=reader+is+super+awesome'
+ *
+ * @param {Object} options the options object
+ * @param {string} options.uri the base uri to modify and add a query to
+ * @param {string} options.search the search term
+ * @param {string} [options.queryKey = s] the key to place in the url.  defaults to s
+ *
+ * @returns {string} The built search url
+ */
+export const buildSearchUrl = ( { uri, search, queryKey = 's' } ) => {
+	const parsedUrl = pick(
+		url.parse( uri, true ),
+		'pathname',
+		'hash',
+		'query',
+	);
+
+	if ( search ) {
+		parsedUrl.query[ queryKey ] = search;
+	} else {
+		delete parsedUrl.query[ queryKey ];
+	}
 
-	return class extends React.Component {
-		state = {
-			searchOpen: false
-		};
+	return url.format( parsedUrl ).replace( /\%20/g, '+' );
+};
 
-		displayName = 'Searchable' + componentName;
+const UrlSearch = Component => class extends React.Component {
+	static displayName = `UrlSearch(${ Component.displayName || Component.name || '' })`;
+	static defaultProps = {
+		search: '',
+		queryKey: 's',
+	}
+
+	state = {
+		searchOpen: false
+	};
 
-		constructor(props) {
-			super(props);
+	componentWillReceiveProps( { search } ) {
+		return ! search && this.setState( { searchOpen: false } );
+	}
 
-			this.doSearch = doSearch.bind( this );
-			this.getSearchOpen = getSearchOpen.bind( this );
+	doSearch = ( query ) => {
+		this.setState( {
+			searchOpen: ( false !== query )
+		} );
+
+		if ( this.onSearch ) {
+			this.onSearch( query );
+			return;
 		}
 
-		componentWillReceiveProps = ( { search } ) => ! search && this.setState( { searchOpen: false } );
+		const searchURL = buildSearchUrl( {
+			uri: window.location.href,
+			search: query,
+			queryKey: this.props.queryKey
+		} );
 
-		render() {
-			return <Component { ...{
-				...this.props,
-				doSearch: this.doSearch,
-				getSearchOpen: this.getSearch,
-			} } />
+		debug( 'search for:', query );
+		if ( this.props.search && query ) {
+			debug( 'replacing URL: ' + searchURL );
```

Comment:
little niggle, but I prefer this form: `debug( 'replacing URL: %s', searchURL )` because it avoids the string concat when debug is disabled